### PR TITLE
Favorite refactoring [READY]

### PR DESCRIPTION
**Pull request by cybrox**
### Stuff to do
- [x] Refactor ember favorite handling (polymorphic)
- [x] Refactor user controller for anime and manga favorites
- [x] Add global component for favorite box of any media type
- [x] Implement drag and drop sort
- [x] Delete favorite records via ember data
- [x] Bulk-update `fav_rank` property of favorites
- [x] Fuzzy search anime and manga input to add media from fav box
- [x] Build new polymorphic relation in ember for new favorites (*undocumented ember shizzle*)
- [x] Add endpoint to effectively add media to your favorites
 - Anime and Manga is using id as identifier for their polymorphic relation on rails side, while the ember application is using slugs as ids.
- [x] Make it look pretty
- [x] Add rails side tests for favorite handling

=====
- [x] Fix bug that allows adding the same title multiple times
- [x] Fix bug that allows adding manga to fav-anime and vice versa
- [x] Fix bug that makes fav container go nuts if search is filled without existing favs
- [x] Fix card hover for search input

=====
- [x] Add favorite button to manga pages
 - This is optional, since these buttons might get removed
- [x] Display editing even if user has no favorites since we can also add favorites now
- [x] Add second container for fav anime for **PRO** users

### CODE REVIEW

File: app/controllers/favorites_controller.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
+class FavoritesController < ApplicationController
+  
+  def index
+    not_found! if params[:user_id].nil? || params[:type].nil?
+
+    @favs = User.find(params[:user_id]).favorites
```
Comment (vikhyat) [inline app/controllers/favorites_controller.rb]: This doesn't need to be an instance variable does it?

File: app/controllers/search_controller.rb [hunk 2]
```diff
@@ -64,10 +64,10 @@ def search_mixed
     users = User.search(@query).page(1).per(20) if users.length == 0
 
     formattedAnime = anime.map { |x|
-      {:type => 'anime', :title => x.title, :image => x.poster_image_thumb, :link => "/anime/#{x.slug}" }
+      {:type => 'anime', :slug => x.slug, :title => x.title, :image => x.poster_image_thumb, :link => "/anime/#{x.slug}" }
```
Comment (vikhyat) [inline app/controllers/search_controller.rb]: We should be using the newer style `{a: b}` hash syntax.

File: app/frontend/app/controllers/user/index.js [hunk 3]
```diff
@@ -79,43 +101,47 @@ export default Ember.ArrayController.extend(HasCurrentUser, {
...
-      // FIXME this should be using ember-data
-      return Ember.$.ajax({
-        url: url,
-        data: {
-          data: data
-        },
-        method: 'POST',
-        error: function () {
-          return console.log("Failed to Update Favorites Ranks");
+      this.store.filter('favorite', function(fav){
```
Comment (vikhyat) [inline app/frontend/app/controllers/user/index.js]: Needs a space before the `{` here and in other places.

File: app/frontend/app/controllers/user/index.js [hunk 4]
```diff
@@ -124,15 +150,23 @@ export default Ember.ArrayController.extend(HasCurrentUser, {
       this.get('user').set('waifu', character.value);
       return this.get('user').set('waifuCharId', character.char_id);
     },
-    loadMoreFavorite_animes: function () {
-      var page;
-      page = this.get('favorite_anime_page');
-      if (page * 6 + 1 <= this.get('favorite_anime').length) {
+
+    favoriteAnimeLoadMore: function () {
```
Comment (vikhyat) [inline app/frontend/app/controllers/user/index.js]: Action names should be verbs, `loadMoreFavoriteAnime`?

File: app/frontend/app/models/media.js [hunk 5]
```diff
@@ -1,5 +1,5 @@
 import DS from 'ember-data';
 
 export default DS.Model.extend({
-  updatedAt: DS.attr('date')
+  updatedAt: DS.attr('date'),
```
Comment (vikhyat) [inline app/frontend/app/models/media.js]: :fallen_leaf:

File: app/frontend/app/templates/components/favorite-container.hbs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  {{#if isEditing}}
+    <div class="grid-list clearfix">
+      {{favorite-search action="submitSearch" currentUser=currentUser}}
+    </div>
+  {{/if}}
+{{else}}
+    <div class="no-favorites text-center">
+      {{#if viewingSelf}}
+        <h3>You have no favorites.</h3>
+        <a href="https://forums.hummingbird.me/t/how-to-adding-an-anime-to-your-favorites"> How to add Favorites</a>
```
Comment (vikhyat) [inline app/frontend/app/templates/components/favorite-container.hbs]: There's an extra space before "How".

File: app/frontend/app/controllers/user/index.js [hunk 7]
```diff
@@ -61,17 +55,45 @@ export default Ember.ArrayController.extend(HasCurrentUser, {
...
+  favorite_manga_list: function() {
+    var manga = this.get('favorite_manga_data'),
+        page  = this.get('favorite_manga_page');
+
+    return manga.slice(0, page * FAVS_PER_PAGE);
+  }.property('favorite_manga_data.length', 'favorite_manga_page'),
 
-    // if using the goPrev and goNext page style, slice the array into a chunk
-    // animes = animes.slice( (page - 1) * 6, page * 6);
+  favorite_anime_load_more: function () {
```
Comment (vikhyat) [inline app/frontend/app/controllers/user/index.js]: Camel-case this and the other stuff.

**Comment by vikhyat** [comment]
Can we add some tests at least for the backend stuff?

**Comment by cybrox** [comment]
Merged in all the new stuff from master to prevent further merge conflicts caused by my first, hacky ember-cli update in this branch as a temporary workaround.
This PR should now be able to be tested with updated ember-cli and merged without any conflicts.

**Pull request merged**